### ISSUE TITLE
Single EXE with MULTIPLE encryptions packages, including SQLCipher 3, 4 and SEE

### ISSUE THREAD

**Issue by mtissington** [sqlitebrowser/sqlitebrowser#1646]
I'm starting a different issue for this ... I think there maybe a better way to support encryption without the need to ship different exe's in the installer.

It is possible to ship a single exe that will support the latest version of sqlite (not encrypted) and a variety of other encryption packages - all from a single exe.

Each sqlite.dll (including sqlcipher3, sqlcipher4) would be installed as a feature on the clients computer.
The user would set their preferred engine in the preferences.

If  the user is using SEE they would specify the path to the SEE dll.

Implementing this in DB Browser would possibly mean using delay load when linking the exe.
And at runtime using the API LoadLibrary/FreeLibrary
[dynamic linking](https://docs.microsoft.com/en-us/windows/desktop/Dlls/using-run-time-dynamic-linking) 

With some AI DB Browser should be able to figure out which library to use.

**Comment by chrisjlocke** [comment]
That might work for Windows.  How would that work for Linux and OSX?  We have to be compatible with a variety of operating systems.

**Comment by mtissington** [comment]
On Linux/OSX you can use dlopen ...

**Comment by chrisjlocke** [comment]
There would have to be multiple dialogs as well though.  SQLCipher wants a password and page size.  Another encryption engine may require additional details.  Another engine may want just a password and ignores the page size.  Supporting that would be .... interesting.  Although it is a logistical nightmare supporting multiple applications.

**Comment by mtissington** [comment]
Well, you already have the dialogs for SQLCipher - they won't change - so that makes support both versions of sqlcipher and unencrypted straightforward ;) ... and then add SEE later according to demand.

**Comment by justinclift** [comment]
People have asked us this before - mostly for Windows - and thus far the answer has been "probably not possible".  However, these days the team has expanded out to include more people with Windows development experience, so it **might** be feasible.

As a data point, when we compile DB4S (either SQLite or SQLCipher) AFAIK the resulting .exe has hard coded info it it which points to the internals pieces of the sqlite.dll or sqlcipher.dll it's paired with.

That's why we're shipping the exe+dll pieces.  If there's some stable + reliable way to update the pointer bits in the .exe prior to a launch, then it might be feasible to have just a single exe.  It kind of sounds like there'd need to be some utility to do this (another exe itself?) which might kind of kill the whole point.  Not sure.

I also suspect (no real idea though) that virus scanners on windows may take a real dim view of applications adjusting things like that.  But, that's a hunch more than based on any first hand knowledge of virus scanner internals. :smile:

**Comment by justinclift** [comment]
> How would that work for Linux and OSX? We have to be compatible with a variety of operating systems.

This is probably just packaging tech we're discussing here, without much (any?) changes to the main DB4S code itself.  Our Linux and macOS installers use completely different approaches anyway, so lets not worry about trying to maintain feature parity in the installers.

* With the Linux packaging, it's generally done by each Linux distribution's packaging system.
* With the macOS packaging... since I'm doing it :wink: I'm thinking to just continue shipping only the SQLCipher version for now.  Same as the previous releases.  No real change in procedure there. :smile:

**Comment by justinclift** [comment]
No worries. :smile:

**Comment by chrisjlocke** [comment]
>  i'm not sure what you mean about hard coded entry points. 

When you ask a DLL to do something, eg, open a database, or decrypt something, you have to explicitly call a method with certain parameters, eg, openDatabase(name_of_database), which may return a value - 1 if it worked, or 0 if it failed.  You 'hard code' that in the code for DB4S under a button or menu option ('Open Datatabase' button for example).  Now if that DLL changes, the new DLL may not have a method called 'openDatabase' but 'DecryptDatabase' and it uses two parameters, not one, and doesn't return a value.  This mismatch is hard to code around neatly.  Do-able, but not easy.

**Comment by karim** [comment]
I think Michael is talking about not linking either SQLite or SQLCipher to DB4S and load them at runtime instead.

So instead of calling `sqlite3_*` functions directly in the code (cause they are linked in the .exe), we use LoadLibrary/GetProcAddress/FreeLibrary. The first function to load sqlite3.dll or sqlcipher.dll, the second to get the address of the function we want inside the dll and then call it, and the last when we are done with it. Then ship sqlite3.dll, sqlcihper3, and sqlcipher4.dll with a single .exe.

This is a very interesting idea, and other OSes can load runtime modules too, so this isn't related to packaging or Windows Installer, and we can remove their labels.

This will require code changes to DB4S and might raise a few questions.

When opening a database file, how do we know which DLL to load that can open this file? We have to load one first before attempting to open the database file. If it failed, does that mean we loaded the wrong DLL or the file isn't a database?

How to handle situations if a database file is opened, using sqlite3.dll for example, then midway the user decided to encrypt it? Do we unload sqlite3.dll and then load sqlcipher4.dll? I don't know how would that work. Doesn't SQLCipher require that database files be created/opened with it first?

**Comment by mtissington** [comment]
That's exactly what I'm talking about - thanks for the clearer explanation :)

Hmm, if the exe is responsible for getting GetProcAddress, then there should be no reason to have both sqlite3 and sqlcipher loaded at the same time (i believe ;)) 

If the users decides to encrypt half way thru, we just call the relevant function and encrypt and switch to using sqlcipher functions as as priority.

**Comment by justinclift** [comment]
Ahhh, now I'm getting what you're both meaning.  Cool. :smile:

I guess the next logical question would be "is there a way to unload the version of sqlite.dll or sqlcipher.dll we've loaded?" so we can then "load the other one" for when we need to switch?

**Comment by mtissington** [comment]
If the exe is keeping track of the function pointers in each dll its technically possible to have them all loaded at the same time! I'm not suggesting this.

Personally I think at the start of a session the users decides which dll to use.
sqlcipher gives you the option to create a new plaintext or unencrypted database.

Leave any AI until after basic functionality is working.

**Comment by justinclift** [comment]
"AI"?

**Comment by mtissington** [comment]
artificial intelligence - ie - make it do what we think the customer wants ..

**Comment by justinclift** [comment]
Not really sure we'd want to add a dependency on GPU hardware, training data, etc just for detecting if a file is a SQLite database? :grin:

**Comment by mtissington** [comment]
as an fyi with this - I have built a sqlcipher 4 dll with visual studio and then dropped in the sqlcipher.dll (replacing the shipped one) and everything seems to still work well.

I built sqlcipher 4 following their instructions for generating the amalgamation.

**Comment by chrisjlocke** [comment]
> and everything seems to still work well

What did you try?  Create a new database, encrypt and re-open it?  Did you try opening an-already encrypted database (encrypted with an older version of SQLCipher)?

**Comment by mtissington** [comment]
I tried creating a new database, encrypting it and then reopening it.
I had to change the page size to 4096.
I've not tried to open an v3 sqlcipher database, not sure I have one - but it should be possible if you pragma specify the various cipher options first

**Comment by utelle** [comment]
Maybe using the [SQLite encryption extension](https://github.com/utelle/wxsqlite3/tree/master/sqlite3secure) coming with [wxSQLite3](https://github.com/utelle/wxsqlite3) would be an option to support multiple encryption schemes within a single EXE.

**wxSQLite3** supports several different encryption schemes (among them _[SQLCipher](https://github.com/sqlcipher/sqlcipher) format 1, 2, 3, and 4_ as well as _[sqleet](https://github.com/resilar/sqleet)_) selectable at runtime by using URI parameters or SQL statements on opening a database.

However, the [original SEE](https://www.sqlite.org/see/), the commercial encryption extension offered by the SQLite development team, is not supported, since _SEE_ is not open source.

**Comment by justinclift** [comment]
@utelle Interesting.  With your SQLite encryption extension, is it feasible to use it without wxWidgets?

If we could someone get it working with Qt, that might open up a few possibilities. :smile:

**Comment by utelle** [comment]
@justinclift 
> Interesting. With your SQLite encryption extension, is it feasible to use it without wxWidgets?

Yes, absolutely. The code of the [wxSQLite3 SQLite encryption extension](https://github.com/utelle/wxsqlite3/tree/master/sqlite3secure) is completely independent of **wxWidgets**. In fact, it is a drop-in replacement for **SQLite**, just as **SQLCipher** is. It is based on the _unmodified_ SQLite amalgamation source code. Therefore, upgrading to a new SQLite version is usually very simple and straight forward.

> If we could somehow get it working with Qt, that might open up a few possibilities. :smile:

The [wxSQLite3 SQLite encryption extension](https://github.com/utelle/wxsqlite3/tree/master/sqlite3secure) uses the same C API for encryption as **SQLCipher** does. That is, it implements the functions `sqlite3_key` and `sqlite3_rekey` (resp `sqlite3_key_v2` and `sqlite3_rekey_v2`). Additionally, a few SQL functions are defined to query and set encryption parameters. However, it does not implement additional PRAGMA commands as **SQLCipher**, since this would require changes to the original SQLite code. Since version 4.3.0 the preferred method to set encryption parameters is via the database URI on opening a database.

In short, it should be no big deal to use it together with **sqlitebrowser**. Only the way how to select a certain cipher with certain encryption parameters would have to be added/changed.

**Comment by justinclift** [comment]
Thanks @utelle, that's good info. :smile:

**Comment by AndreasMartin72** [comment]
Supporting **sqleet** (as the  wxSQLite3 SQLite encryption extension does) would be a big deal to me! I've many databases based on **sqleet**, which I then may easily explore with **sqlitebrowser**, too.

**Comment by bernd-wechner** [comment]
Yes, a shame this is still open and not closed completed ;-). I have a sqleet database and would love to browse it with a cool GUI like sqlitebrowser's.

**Comment by m4heshd** [comment]
Any chance of this happening?

**Comment by justinclift** [comment]
@m4heshd Probably not in the near future, unless someone gets enthusiastic about it and decides they want to put in the effort to make it happen. :smile:

**Comment by m4heshd** [comment]
Hahaaa. Actually I'm willing to give it a try once sqleet moves forward from the wall it's been hit at the moment. Sqlcipher still works right?

**Comment by justinclift** [comment]
Yep, SQLCipher works fine. :smile:
